New RZIL integration

Heersin opened the pull request:
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

1. [New IL](https://github.com/BinaryAnalysisPlatform/bap/pull/960) as an independent library which relies on `rz_util` only.
2. IL in core analysis are moved to `core/cil.c`.  (Related pr of moving esil -> #1360 )
3. Change the tech of collecting `stats` and `trace` in the old ESIL
4. Fix some bugs in the old pr #1361 #1451

**Test plan**

Should pass the old rizin tests.

**Closing issues**

closes #277 

**More Info**
0. Use tree-like approach
1. This pr created for `asan build` to detect bug.
2. Only enabled in `analysis_bf.c`. 
2. `Register arena` changes in other module (such as `debug`) are not covered in current new IL.

Heersin committed:
* Roughly integrate with New IL

Heersin committed:
Fix code according to reviewer comments

 * Fix db/cmd/esil tests and anlysis_bf bugs

 * Change BUILD_ header to RZ header

 * Fix macros and types

 * Remove unnecessary string and bit types and unused code

 * Add autolabeler

 * Various file name and function name changes

 * Fix broken test

 * Add unit test for il_definitions

 * Change the index behaviour of bitvector

Heersin committed:
Add some test about bitvector

- Fix some comments

- Add ut8 ut16 convertions

Heersin committed:
Address new comments

* Rename functions

* Add as_string and remove some unused debug functions

* Add doxygen docs

* Small fix

* Fix header protector to corresponding name

* Add null checks and RZ_NONNULL

* Add more unit tests of rzil_vm, effect and mem

* Minimal migration from index approach tree-like approach

* Fix bug in new rzil op struct

Heersin commented:
The `trace` and `stats` techs of the old index approach replies on temporary value. Now it has been removed. We need to implement the `trace`/`stats` in the rzil. It aims to collect info reg/mem read/write.

Currently, the API of opcode evaluation is :
rzil_parse_root(RzILVM *vm, RzILOp *op);

it's a recursive function to evaluate the expression. so we may add an extra arg as Callback function
rzil_parse_root(RzILVM *vm, RzILOp *op, ParseCallbacks f);

the callback function ccould be 
```C
rzil_focus_trace (RzILOp *op)
 - mem read
 - mem write
 - reg read
 - reg write
```
etc.

but what I worry about is the possible header conflict. 
Since our new trace info structure defined in higher level of rizin project while the rzil can only rely on `rz_util`

we may not be able to directly use the function and trace info struct in rz_analysis.h

1. Defined a temporary struct to store the collected info ?
2. Implement the similar parse/emulate function in the same level or higher one. As esil did.

That's two possible solutions I have for now. Do you have any idea ?

XVilka commented:
@Heersin I think adding some new structure is a way to go. Not sure what @ret2libc @thestr4ng3r @wargio think.

wargio commented on librz/core/analysis_tp.c:
```diff
@@ -9,14 +9,17 @@
+
 	core->dbg->trace = rz_debug_trace_new();
 	core->analysis->esil->trace = rz_analysis_esil_trace_new(core->analysis->esil);
+	core->analysis->rzil->trace = rz_analysis_rzil_trace_new(core->analysis, core->analysis->rzil);
```
if you have `core->analysis` why passing `core->analysis->rzil` ?

Heersin replied:
Yes it passes these 2 because I followed the ESIL api. We can only maintain one argument `core->analysis`.

XVilka replied:
Yes, lets skip and use just `core->analysis` for both.

wargio commented on librz/core/cil.c:
```diff
@@ -428,3 +428,53 @@ RZ_IPI void rz_core_analysis_esil_default(RzCore *core) {
+
+	// try load from vm
+	// fetch and parse if no opcode
+	ut8 code[32];
+	// analysis current data to trigger rzil_set_op_code
+	(void)rz_io_read_at_mapped(core->io, addr, code, sizeof(code));
```
since you are ignoring the rz_io_read_at_mapped return value, maybe it should be better to initialize code to 0 or any other value that represent NOP or invalid op.

XVilka replied:
I recommend not to ignore the value and do the function exit on the error

wargio commented on librz/il/definitions/bag.c:
```diff
@@ -0,0 +1,137 @@
+ * \return RzILBag pointer
+ */
+RZ_API RzILBag *rz_il_new_bag(int capacity, RzILBagFreeFunc func) {
+	RzILBag *bag = RZ_NEW0(struct rz_il_bag_t);
```
```suggestion
	RzILBag *bag = RZ_NEW0(RzILBag);
```

wargio commented on librz/il/definitions/bool.c:
```diff
@@ -0,0 +1,79 @@
+ * \return bool RzILBool, pointer to bool value
+ */
+RZ_API RzILBool *rz_il_new_bool(bool true_or_false) {
+	RzILBool *ret = (RzILBool *)malloc(sizeof(struct bool_t));
```
```suggestion
	RzILBool *ret = RZ_NEW(RzILBool);
```

wargio commented on librz/il/definitions/effect.c:
```diff
@@ -0,0 +1,213 @@
+ */
+RZ_API RzILDataEffect *rz_il_effect_new_data(void) {
+	RzILDataEffect *ret;
+	ret = (RzILDataEffect *)malloc(sizeof(struct rzil_data_effect_t));
```
```suggestion
	ret = RZ_NEW(RzILDataEffect);
```

wargio commented on librz/il/definitions/effect.c:
```diff
@@ -0,0 +1,213 @@
+ */
+RZ_API RzILCtrlEffect *rz_il_effect_new_ctrl(void) {
+	RzILCtrlEffect *ret;
+	ret = (RzILCtrlEffect *)malloc(sizeof(struct rzil_control_effect_t));
```
```suggestion
	ret = RZ_NEW(RzILCtrlEffect);
```

wargio commented on librz/il/definitions/effect.c:
```diff
@@ -0,0 +1,213 @@
+ */
+RZ_API RzILEffect *rz_il_wrap_ctrl_effect(RzILCtrlEffect *eff) {
+	RzILEffect *ret;
+	ret = (RzILEffect *)malloc(sizeof(struct rzil_effect_union_t));
```
Why 2 different types?

Heersin replied:
You mean the `ctrl` one and the `data` one? Because there are 2 different effect types defined in core theory, they may carry different types of info.

wargio commented on librz/il/definitions/effect.c:
```diff
@@ -0,0 +1,213 @@
+ */
+RZ_API RzILEffect *rz_il_wrap_data_effect(RzILDataEffect *eff) {
+	RzILEffect *ret;
+	ret = (RzILEffect *)malloc(sizeof(struct rzil_effect_union_t));
```
same as above

wargio commented on librz/il/definitions/effect.c:
```diff
@@ -0,0 +1,213 @@
+		ret = rz_il_wrap_data_effect(rz_il_effect_new_data());
+		break;
+	case EFFECT_TYPE_NON:
+		ret = (RzILEffect *)malloc(sizeof(struct rzil_effect_union_t));
```
same here

wargio commented on librz/il/definitions/effect.c:
```diff
@@ -0,0 +1,213 @@
+ * \return Pointer to label
+ */
+RZ_API RzILEffectLabel *rz_il_effect_new_label(RZ_NONNULL const char *name, EFFECT_LABEL_TYPE type) {
+	RzILEffectLabel *lbl = (RzILEffectLabel *)RZ_NEW0(struct rzil_effect_label_t);
```
```suggestion
	RzILEffectLabel *lbl = RZ_NEW0(RzILEffectLabel);
```

wargio commented on librz/il/definitions/mem.c:
```diff
@@ -0,0 +1,79 @@
+ * \return RzILMem
+ */
+RZ_API RzILMem rz_il_new_mem(int min_unit_size) {
+	RzILMem ret = (RzILMem)RZ_NEW0(struct rzil_mem_t);
```
```suggestion
	RzILMem ret = RZ_NEW0(RzILMem);
```

wargio commented on librz/il/definitions/mem.c:
```diff
@@ -0,0 +1,79 @@
+ * \param min_unit_size, minimal size of a data unit of current arch
+ * \return RzILMem
+ */
+RZ_API RzILMem rz_il_new_mem(int min_unit_size) {
```
Can be negative? isn't better to use unsigned ? like ut32 ?

Heersin replied:
better to be unsigned number, no negative allowed

wargio commented on librz/il/definitions/value.c:
```diff
@@ -0,0 +1,91 @@
+ */
+RZ_API RzILVal *rz_il_new_value(void) {
+	RzILVal *ret;
+	ret = (RzILVal *)malloc(sizeof(struct rz_il_val_t));
```
```suggestion
	ret = RZ_NEW(RzILVal);
```

wargio commented on librz/il/definitions/value.c:
```diff
@@ -0,0 +1,91 @@
+ * \return temp RzILTemp, pointer to the new temporary value
+ */
+RZ_API RzILTemp rz_il_new_temp(void) {
+	RzILTemp temp = RZ_NEW0(struct rz_il_tempv_t);
```
```suggestion
	RzILTemp temp = RZ_NEW0(RzILTemp);
```

wargio commented on librz/il/definitions/variable.c:
```diff
@@ -0,0 +1,36 @@
+RZ_API RzILVar *rz_il_new_variable(const char *name) {
+	RzILVar *ret;
+
+	ret = (RzILVar *)malloc(sizeof(struct rz_il_var_t));
```
```suggestion
	ret = RZ_NEW(RzILVar);
```

wargio commented on librz/il/rzil_opcodes.c:
```diff
@@ -0,0 +1,176 @@
+#include <rz_il/rzil_opcodes.h>
+
+RzILOp *rz_il_new_empty_op(void) {
+	RzILOp *ret = (RzILOp *)malloc(sizeof(struct RzILOp_t));
```
```suggestion
	RzILOp *ret = RZ_NEW(RzILOp);
```

wargio commented on librz/il/rzil_opcodes.c:
```diff
@@ -0,0 +1,176 @@
+ * \return RzILOp, a pointer to an empty opcode instance
+ */
+RZ_API RzILOp *rz_il_new_op(RzILOPCode code) {
+	RzILOp *ret = (RzILOp *)malloc(sizeof(struct RzILOp_t));
```
```suggestion
	RzILOp *ret = RZ_NEW(RzILOp);
```

wargio commented on librz/il/rzil_vm.c:
```diff
@@ -0,0 +1,462 @@
+// WARN : convertion breaks the original data
+static RzILBool *bitv_to_bool(RzILBitVector *bitv) {
+	RzILBool *result;
+	if (rz_il_bv_is_zero_vector(bitv)) {
+		result = rz_il_new_bool(false);
+	} else {
+		result = rz_il_new_bool(true);
+	}
```
```suggestion
	result = rz_il_new_bool(!rz_il_bv_is_zero_vector(bitv));
```

wargio commented on librz/il/theory_effect.c:
```diff
@@ -0,0 +1,149 @@
+	RzILOpBranch *op_branch = op->op.branch;
+
+	RzILBool *condition = rz_il_evaluate_bool(vm, op_branch->condition, type);
+	RzILEffect *true_branch, *false_branch;
+	RzILEffect *ret;
+
+	if (condition->b) {
+		true_branch = (op_branch->true_eff == NULL) ? rz_il_effect_new(EFFECT_TYPE_NON) : rz_il_evaluate_effect(vm, op_branch->true_eff, type);
+		ret = true_branch;
+	} else {
+		false_branch = (op_branch->false_eff == NULL) ? rz_il_effect_new(EFFECT_TYPE_NON) : rz_il_evaluate_effect(vm, op_branch->false_eff, type);
+		ret = false_branch;
+	}
```
```suggestion
	RzILEffect *ret;
	if (condition->b) {
		// true branch
		ret = (op_branch->true_eff == NULL) ? rz_il_effect_new(EFFECT_TYPE_NON) : rz_il_evaluate_effect(vm, op_branch->true_eff, type);
	} else {
		// false branch
		ret = (op_branch->false_eff == NULL) ? rz_il_effect_new(EFFECT_TYPE_NON) : rz_il_evaluate_effect(vm, op_branch->false_eff, type);
	}
```

wargio commented on librz/il/theory_init.c:
```diff
@@ -0,0 +1,37 @@
+	RzILBool *condition = rz_il_evaluate_bool(vm, op_ite->condition, type);
+	RzILVal *ret;
+
+	if (condition->b) {
+		*type = RZIL_OP_ARG_VAL;
+		ret = rz_il_evaluate_val(vm, op_ite->x, type); // true branch
+	} else {
+		*type = RZIL_OP_ARG_VAL;
+		ret = rz_il_evaluate_val(vm, op_ite->y, type); // false branch
+	}
```
```suggestion
	*type = RZIL_OP_ARG_VAL;
	if (condition->b) {
		ret = rz_il_evaluate_val(vm, op_ite->x, type); // true branch
	} else {
		ret = rz_il_evaluate_val(vm, op_ite->y, type); // false branch
	}
```

wargio commented on librz/il/vm_layer.c:
```diff
@@ -0,0 +1,353 @@
+	vm->addr_size = addr_size;
+	vm->data_size = data_size;
+
+	vm->vm_global_variable_list = (RzILVar **)calloc(RZ_IL_VM_MAX_VAR, sizeof(RzILVar *));
```
```suggestion
	vm->vm_global_variable_list = RZ_NEWS0(RzILVar *, RZ_IL_VM_MAX_VAR);
```

wargio commented on librz/il/vm_layer.c:
```diff
@@ -0,0 +1,353 @@
+	vm->ct_opcodes = ht_pp_new_opt(&ops_options);
+
+	// init jump table of labels
+	vm->op_handler_table = (RzILOpHandler *)malloc(sizeof(RzILOpHandler) * RZIL_OP_MAX);
+	memset(vm->op_handler_table, 0, RZIL_OP_MAX);
```
```suggestion
	vm->op_handler_table = RZ_NEWS0(RzILOpHandler, RZIL_OP_MAX);
```

wargio commented on librz/include/rz_analysis.h:
```diff
@@ -807,6 +809,13 @@ typedef enum rz_analysis_data_type_t {
 } RzAnalysisDataType;
 
+// For switchting to AST-like approach in the future
+typedef struct rz_analysis_rzil_op_t {
+	RzPVector *ops;
+	// TODO : use root approach in the future
+	RzILOp **root_node;
+} RzAnalysisRzilOp;
+
```
This is just a comment:
After we merge this, we should move esil and rz_il to it's own headers.

XVilka replied:
This comment should be removed then, since it's already root approach.

wargio reviewed (changes requested):
looks good with these new changes.

ret2libc commented:
> The `trace` and `stats` techs of the old index approach replies on temporary value. Now it has been removed. We need to implement the `trace`/`stats` in the rzil. It aims to collect info reg/mem read/write.
> 
> Currently, the API of opcode evaluation is :
> rzil_parse_root(RzILVM *vm, RzILOp *op);
> 
> it's a recursive function to evaluate the expression. so we may add an extra arg as Callback function
> rzil_parse_root(RzILVM *vm, RzILOp *op, ParseCallbacks f);

I'm not 100% sure I understood the problem, but I think having a callback that takes the ILOp as argument (maybe also the RzILVM as well) should be enough to support all kinds of traversal of the OPs without putting anything specific in the RzIL module.

XVilka commented on librz/il/vm_layer.c:
```diff
@@ -0,0 +1,334 @@
+	rz_il_parse_op_root(vm, root, &type);
+}
+
+RZ_API void rz_il_vm_list_step(RzILVM *vm, RzPVector *op_list) {
```
Missing Doxygen.

XVilka commented on librz/il/rzil_vm.c:
```diff
@@ -0,0 +1,457 @@
+	case RZIL_OP_ARG_EFF:
+	case RZIL_OP_ARG_MEM:
+	default:
+		RZ_LOG_ERROR("BAD TYPE DETECTED\n");
```
Type of what? Write something like "RZIL: Bad argument type"

XVilka commented on librz/il/rzil_vm.c:
```diff
@@ -0,0 +1,457 @@
+	case RZIL_OP_ARG_VAL:
+	case RZIL_OP_ARG_MEM:
+	default:
+		RZ_LOG_ERROR("BAD TYPE DETECTED\n");
```
Same

XVilka commented on librz/il/rzil_vm.c:
```diff
@@ -0,0 +1,457 @@
+	case RZIL_OP_ARG_EFF:
+	case RZIL_OP_ARG_MEM:
+	default:
+		RZ_LOG_ERROR("BAD TYPE DETECTED\n");
```
Same

XVilka commented on librz/il/rzil_vm.c:
```diff
@@ -0,0 +1,457 @@
+	case RZIL_OP_ARG_EFF:
+	case RZIL_OP_ARG_MEM:
+	default:
+		RZ_LOG_ERROR("BAD TYPE DETECTED\n");
```
Same

XVilka reviewed (changes requested):
Overall in a good shape now. Only a few minor nitpicks left, and should be ready to merge. We can improve the implementation further after merging, IMHO.

Heersin committed:
Add better error message

XVilka commented:
@Heersin could you please rebase on the latest `dev` and fix some unaddressed comments - there is a conflict now?

Heersin committed:
Add doxygen docs

thestr4ng3r commented on librz/core/cmd_analysis.c:
```diff
@@ -5325,6 +5399,10 @@ static void cmd_analysis_esil(RzCore *core, const char *input) {
 		break;
 	}
+	case 'z': { // "aez"
+		cmd_analysis_rzil(core, input + 1);
```
Was anything already decided whether the `aez` commands should be implemented in newshell directly? @ret2libc @XVilka

XVilka replied:
@thestr4ng3r I think it's fine to have it as an oldshell in this PR to avoid irrelevant delays. They can be converted later in https://github.com/rizinorg/rizin/issues/1533

thestr4ng3r replied:
ack

thestr4ng3r commented on librz/include/rz_il/rzil_opcodes.h:
```diff
@@ -0,0 +1,447 @@
+ */
+struct rzil_op_int_t {
+	ut32 length; ///< s -- sort(type), length of bitvector
+	int value; ///< x -- value of bitvector
```
Shouldn't this rather just contain a single `RzILBitVector` and be called something like `RzILOpBitVector`?

thestr4ng3r commented on librz/include/rz_il/rzil_opcodes.h:
```diff
@@ -0,0 +1,447 @@
+
+	// Memory
+	RZIL_OP_LOAD,
+	RZIL_OP_STORE,
```
Wouldn't `RZIL_OP_LOAD` actually be a bit vector op and `RZIL_OP_STORE` be an effect?

XVilka replied:
No, the operand and effect are separated in this case, see the original implementation: 
- http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/Effect/Sort/index.html

thestr4ng3r replied:
Interesting, where are the operands in this link?

thestr4ng3r commented on librz/include/rz_il/rzil_opcodes.h:
```diff
@@ -0,0 +1,447 @@
+ *  [LSB] lsb x is the least significant bit of x.
+ */
+struct rzil_op_msb_lsb_t {
+	RzIL_BITV void *bv; ///< index of bitvector operand
```
Why are all of these types only `void *`? It appears to me like they could be just `RzILOp *`.

On another note, have you considered splitting up `RzILOp` into multiple distinct structs depending on their return types?
So you would have `RzILOpBV`, `RzILOpBool` and `RzILOpEffect`, each with a union inside covering only the cases for the respective type.

Heersin replied:
@thestr4ng3r  Oh the separating looks like a good idea, I will consider that. As for the first question, the reason to use `void*` is `RzILOp` defined in another file and relies on different `rzil_op_*_t`, and thus I cannot use `RzILOp` inside `rzil_op_*_t`. 

I used `typedef struct rzil_op_msb_lsb_t* RzILOpMsb` to bypass this limitation. Do you have better idea ?

Heersin replied:
I have found a way now.

ret2libc commented on librz/analysis/rzil/rzil.c:
```diff
@@ -0,0 +1,144 @@
+	if (!rzil) {
+		return NULL;
+	}
+	rzil->vm = RZ_NEW0(struct rz_il_vm_t);
```
```	rzil->vm = RZ_NEW0(RzILVM);```

XVilka replied:
@Heersin you forgot also this.

ret2libc commented on librz/analysis/rzil/rzil.c:
```diff
@@ -0,0 +1,144 @@
+ * \param nonull int is pc cannot be null ?
+ * \return true if setup, else return false
+ */
+RZ_API bool rz_analysis_rzil_setup(RzAnalysis *analysis) {
```
Is this expected to be call only after rz_analysis_rzil_cleanup? If so, add a `rz_return_val_if_fail(!analysis->rzil, false)` to make sure that we don't leak the previous value of `analysis->rzil`.

Heersin replied:
Call at the beginning or after the cleanup. Great, I will add one.

ret2libc commented on librz/analysis/rzil/rzil.c:
```diff
@@ -0,0 +1,144 @@
+}
+
+/**
+ * Set operand list starting at the specified address for the RZIL VM
```
What does it mean to "set the operand list"? Sorry, I don't get it

Heersin replied:
It's used to set the opcode to an address, likes a cache. We don't cache things in current stage. So I will remove it and the bad comments. Sorry for the noising.

ret2libc commented on librz/core/cil.c:
```diff
@@ -428,3 +428,53 @@ RZ_IPI void rz_core_analysis_esil_default(RzCore *core) {
 	rz_core_seek(core, at, true);
 }
+
+RZ_IPI void core_rzil_init(RzCore *core) {
```
Is this really needed? I think it is used only in this file and you can probably just directly call `rz_analysis_rzil_setup`.

Heersin replied:
Not needed for now. I put a similar function here as ESIL did. It might be used in the future as a new command. We can remove it.

ret2libc commented on librz/core/cmd_analysis.c:
```diff
@@ -4805,6 +4805,79 @@ static void __analysis_esil_function(RzCore *core, ut64 addr) {
+	case 's': // "aezs"
+		switch (input[1]) {
+		case '?': // "aez?"
+			rz_core_cmd0(core, "ae?~aes");
```
?? what is this?

Heersin replied:
Since we use the old ESIL and a new one. I try to introduce a new command family for `RZIL`. Though it's used for developer now.

XVilka replied:
This one - better to not print anything or just print static string. It will be addressed during the https://github.com/rizinorg/rizin/issues/1533

ret2libc commented on librz/il/vm_layer.c:
```diff
@@ -0,0 +1,339 @@
+ * Close and clean vm
+ * \param vm RzILVM* pointer to VM
+ */
+RZ_API void rz_il_vm_close(RzILVM *vm) {
```
We don't usually use `close`. Also, there is no `open` AFAIK, so I think it's better to just use `init`/`fini` as usual. And given that the `init` does not allocate the memory for RzILVM, the `fini` should not free the `vm` itself.

ret2libc commented on test/unit/test_il_vm.c:
```diff
@@ -0,0 +1,147 @@
+	rz_il_vm_init(vm, 0, 8, 16);
+
+	// 1. create variables
+	RzILVar *var_r1 = rz_il_vm_create_variable(vm, "r1");
```
How can you assign a type to a `RzILVar`? Also, why is the type only in the `RzILVal` and not in the `RzILVar`? Is it possible this way to assign different `RzILVal` types to the same `RzILVar`?

XVilka replied:
There is a a clear distinction between variables and values, see:
- http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/Value/index.html
- http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/Var/index.html

On the other hand, you are right, the sorts of Vars and Variables are related:
```ocaml
val sort : 'a t -> 'a Value.sort
(*  sort v is variable's sort. *)
```
You can find all sorts of the Value here: http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/Value/Sort/index.html

# Module Value.Sort

Value Sorts.

A concrete and extensible representation of a value sort. The sort usually holds the static information about the value representation, like the width of a bitvector, the representation format of a floating point number, and so on.

This module is mostly needed when a new sort is defined. The Core Theory provides a predefined collection of sorts, here is the list:
- **Bitv** - bitvectors, e.g., BitVec(i)
- **Mem** - memories, e.g., Mem(BitVec(i), BitVec(j)
- **Float** - floating points, e.g., Float(IEEE754(2, 8, 23), BitVec(32);
- **Rmode** - rounding mode, e.g., Rmode.

This module defines a simple DSL for specifying sorts, the DSL grammar is made only from three rules:
```ocaml
 sort = sym | int | sort(sort)
```
The DSL is embedded into the host language with an infix operator @-> for application, e.g., OCaml grammar for sorts is:
```ocaml
v sort = sym exp | int exp | sort "@->" sort exp = ?a valid OCaml expression? v
```
Both symbols and numbers are indexed with a type index, which serves as a witness of the sort value, e.g.,
```ocaml
type int8
let int8 : int8 num sort = Sort.int 8
```
Type indices enable explicit reflection of the target language type system in the host type system, while still keeping the typing rules under designer's control.

ret2libc commented on test/unit/test_il_vm.c:
```diff
@@ -0,0 +1,147 @@
+	rz_il_vm_update_label(vm, "lazy", addr);
+	lazy_addr = rz_il_hash_find_addr_by_lblname(vm, "lazy");
+	is_equal_bv = rz_il_bv_cmp(lazy_addr, addr) == 0 ? true : false;
+	mu_assert("Update lazy label successfully", is_equal_bv);
```
Use `mu_assert_true`, it accepts the argument like all the other functions, with the message as last arg.

XVilka replied:
@Heersin you forgot this one

ret2libc commented on test/unit/test_il_definitions.c:
```diff
@@ -0,0 +1,453 @@
+#include "minunit.h"
+
+static bool is_equal_bv(RzILBitVector *x, RzILBitVector *y) {
+	return (rz_il_bv_cmp(x, y) == 0 ? true : false);
```
`return rz_il_bv_cmp(x, y) == 0;`

XVilka replied:
@Heersin you forgot this one

ret2libc commented on test/unit/test_il_definitions.c:
```diff
@@ -0,0 +1,453 @@
+	rz_il_bv_free(result);
+
+	result = rz_il_bv_or(x, y);
+	printf("\n");
```
remove?

ret2libc commented on librz/include/rz_il/vm_layer.h:
```diff
@@ -0,0 +1,25 @@
+// VM high level operations
+RZ_API bool rz_il_vm_init(RzILVM *vm, ut64 start_addr, int addr_size, int data_size);
+RZ_API void rz_il_vm_close(RzILVM *vm);
+void rz_il_vm_step(RzILVM *vm, RzILOp *root);
```
this should have RZ_API

XVilka replied:
@Heersin you forgot this one

ret2libc commented on librz/il/vm_layer.c:
```diff
@@ -0,0 +1,339 @@
+	return NULL;
+}
+
+void rz_il_vm_step(RzILVM *vm, RzILOp *root) {
```
RZ_API + doxygen doc

XVilka replied:
@Heersin you forgot this one

ret2libc commented on librz/il/rzil_vm.c:
```diff
@@ -0,0 +1,460 @@
+ * \param ... op RzILOp, op will be pushed to vector one by one
+ * \return oplist RzPVector*, pointer to the opcode
+ */
+RZ_API RzPVector *rz_il_make_oplist(int num, ...) {
```
This can be changed later, but I think it's less error prone to have NULL as terminator instead of having `num` as the number of following items.

ret2libc commented on librz/include/rz_il/rzil_vm.h:
```diff
@@ -0,0 +1,129 @@
+	RZIL_OP_ARG_EFF,
+	RZIL_OP_ARG_MEM,
+	RZIL_OP_ARG_INIT
+} RZIL_OP_ARG_TYPE;
```
`RzILOpArgType` no upper case for the enum type.

ret2libc commented on librz/il/rzil_vm.c:
```diff
@@ -0,0 +1,460 @@
+}
+
+// VM auto convertion
+RzILBitVector *rz_il_evaluate_bitv(RzILVM *vm, RzILOp *op, RZIL_OP_ARG_TYPE *type) {
```
can this be explained a bit better in the doc? What does it actually do? Sorry for the maybe stupid question.

Heersin replied:
It evaluate an expression and ensure you to get a bitvector. When evaluating a sub expression, we may get a bool/bv/val value, but the opcode may require a bitvector operand. This function will convert valid types to a bitvector.

ret2libc commented on librz/include/rz_il/rzil_vm.h:
```diff
@@ -0,0 +1,129 @@
+RZ_API RzPVector *rz_il_make_oplist(int num, ...);
+
+// VM auto convertion
+RzILBitVector *rz_il_evaluate_bitv(RzILVM *vm, RzILOp *op, RZIL_OP_ARG_TYPE *type);
```
I think there are no tests for any of these `evaluate` functions, are there? Can they be added or be better explained in doxygen doc?

Heersin replied:
Ok, I will added them

ret2libc commented on librz/include/rz_il/definitions/value.h:
```diff
@@ -0,0 +1,47 @@
+#define RzIL_BOOL
+#define RzIL_VAR
+#define RzIL_EFF
+#define RzIL_LABLE
```
typo here. However, I think these are not necessary anymore given that we can just use the right types now instead of `void *`

Heersin replied:
Yes

ret2libc commented:
There's a lot of work here, nice job. I don't want to be tooo picky here, as otherwise we'll never merge this. I think there are however just few more things to adjust, in particular the  void * thing as mentioned also by @thestr4ng3r .

XVilka commented:
@Heersin could you please rebase your PR and address the feedback?

Heersin commented:
> @Heersin could you please rebase your PR and address the feedback?

Sure, I will review it tonight

Heersin committed:
Address some comments

Address some rename comments

Add test for evaluation

Address some old comments

Fix tests

XVilka reviewed (changes requested):
@Heersin you forgot 6 more feedback comments, please take a look again. Apart from that, it's good to merge, I think. cc @ret2libc @thestr4ng3r

XVilka commented:
Two ASAN failed tests are unrelated to this PR and happen on the latest `dev` as well:
```
[XX] TIMEOUT db/formats/elf/sections more than 65535 segments
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc iSS~? gzip://bins/elf/core/more-than-65535-segments.gz
-- stdout
--- expected
+++ actual
@@ -1,1 +1,0 @@
-131073

-- stderr
Cannot retrieve regstate on: AMD x86-64 architecture (not yet supported)

-- exit status: -1

[XX] TIMEOUT db/formats/mdmp mdmp 64bit - strings
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc 'iz~0x002e2fe1[1,2,3,4,5,6,7,8]
iz~0x002e2feb[1,2,3,4,5,6,7,8]
iz~0x002e2ff5[1,2,3,4,5,6,7,8]
iz~0x00330d85[1,2,3,4,5,6,7,8]
iz~0x003310c1[1,2,3,4,5,6,7,8]
iz~0x00161a37[1,2,3,4,5,6,7,8]
iz~0x00161c7f[1,2,3,4,5,6,7,8]
iz~0x00161dbf[1,2,3,4,5,6,7,8]
iz~0x001d9d45[1,2,3,4,5,6,7,8]
iz~0x001da0a5[1,2,3,4,5,6,7,8]
iz~0x003fad45[1,2,3,4,5,6,7,8]
iz~0x003fb0ad[1,2,3,4,5,6,7,8]
iz~0x00480857[1,2,3,4,5,6,7,8]
iz~0x004808c7[1,2,3,4,5,6,7,8]
iz~0x00499cf5[1,2,3,4,5,6,7,8]
iz~0x0049a045[1,2,3,4,5,6,7,8]
' bins/mdmp/hello64.dmp
-- stdout
--- expected
+++ actual
@@ -8,9 +8,3 @@
 0x00161dbf 0x7759e130 45 92 C:_Windows_System32_kernel32.dll utf16le \REGISTRY\USER\*\SOFTWARE\Classes\Wow6432Node
 0x001d9d45 0x776160b6 15 32 C:_Windows_System32_kernel32.dll utf16le VS_VERSION_INFO
 0x001da0a5 0x77616416 11 24 C:_Windows_System32_kernel32.dll utf16le VarFileInfo
-0x003fad45 0x7fefd5080b6 15 32 C:_Windows_System32_KERNELBASE.dll utf16le VS_VERSION_INFO
-0x003fb0ad 0x7fefd50841e 11 24 C:_Windows_System32_KERNELBASE.dll utf16le VarFileInfo
-0x00480857 0x7fefdb33bc8 4 10 C:_Windows_System32_msvcrt.dll utf16le PATH
-0x004808c7 0x7fefdb33c38 10 22 C:_Windows_System32_msvcrt.dll utf16le SystemRoot
-0x00499cf5 0x7fefdb4d066 15 32 C:_Windows_System32_msvcrt.dll utf16le VS_VERSION_INFO
-0x0049a045 0x7fefdb4d3b6 11 24 C:_Windows_System32_msvcrt.dll utf16le VarFileInfo

-- stderr
[WARN] Invalid or unsupported enumeration encountered 21
[WARN] Invalid or unsupported enumeration encountered 22
[INFO] Parsing data sections for large dumps can take time, please be patient (but if strings ain't your thing try with -z)!

-- exit status: -1
```

XVilka committed:
Merge branch 'dev' into asan_integrate_new_il
